### PR TITLE
Issue 4886: Add support for Azure Blob

### PR DESCRIPTION
**Pull request by amymareemc**
### Describe your changes :

I worked on adding support for Azure Blob for metadata backup dump (See [issue #4868](https://github.com/open-metadata/OpenMetadata/issues/4868). I did this by in `backup.py`: 
- Creating a new parameter in `run_backup` called `uploadDestinationType` which can either be "AWS" or "Azure"
- Depending on specified `uploadDestinationType`, `run_backup` calls` upload_backup_aws` (renamed from `upload_backup`) or the newly added `upload_backup_azure`
- `upload_backup_azure` follows the documentation on adding a blob to a container (see [here](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-python?tabs=managed-identity%2Croles-azure-portal%2Csign-in-azure-cli))
- In `cmd.py`, I also updated backup to take the `uploadDestinationType` and limited the options to either AZURE or AWS using `@click.option`

I am currently unsure on how to test this implementation. Sought advice on slack and was told to submit a draft PR. 

#
### Type of change :

- [ ] Bug fix
- [x] Improvement
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Frontend Preview (Screenshots) :

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: ingestion/src/metadata/cli/backup.py [hunk 1]
```diff
@@ -98,13 +98,56 @@ def upload_backup(endpoint: str, bucket: str, key: str, file: Path) -> None:
...
+        raise err
+
+    click.secho(
+        f"Uploading {file} to {account_url}/{container}...",
+        fg="bright_green",
+    )
+
+    try:
+        # Create a blob client using the local file name as the name for the blob
+        blob_client = blob_service_client.get_blob_client(container=container, blob=file.name)
```
Comment (pmbrull) [inline ingestion/src/metadata/cli/backup.py:126]: Where are we initializing the `blob_service_client` variable?

File: ingestion/src/metadata/cli/backup.py [hunk 2]
```diff
@@ -98,13 +98,56 @@ def upload_backup(endpoint: str, bucket: str, key: str, file: Path) -> None:
...
+
+        # Upload the created file
+        with open(file=file.absolute, mode="rb") as data:
+            blob_client.upload_blob(data)
+        
+
```
Comment (pmbrull) [inline ingestion/src/metadata/cli/backup.py:132]: I think that the formatting is not correct. Please run `make py_format` to make sure that the Python checkstyle passes.

File: ingestion/src/metadata/cli/backup.py [hunk 3]
```diff
@@ -98,13 +98,56 @@ def upload_backup(endpoint: str, bucket: str, key: str, file: Path) -> None:
...
+        f"Uploading {file} to {account_url}/{container}...",
+        fg="bright_green",
+    )
+
+    try:
+        # Create a blob client using the local file name as the name for the blob
+        blob_client = blob_service_client.get_blob_client(container=container, blob=file.name)
+
+        # Upload the created file
+        with open(file=file.absolute, mode="rb") as data:
```
Comment (pmbrull) [inline ingestion/src/metadata/cli/backup.py:129]: here we might be missing the `encoding`. The linting on the CI will fail. You can set `encoding=UTF_8`, we have a constant somewhere you can import

File: ingestion/src/metadata/cli/backup.py [hunk 4]
```diff
@@ -143,5 +186,11 @@ def run_backup(  # pylint: disable=too-many-arguments
     )
 
     if upload:
-        endpoint, bucket, key = upload
-        upload_backup(endpoint, bucket, key, out)
+        if uploadDestinationType == "AWS":
```
Comment (pmbrull) [inline ingestion/src/metadata/cli/backup.py:189]: Here we could have an Enum with `AWS` and `Azure` as possible options to make it easier to know what scenarios we are covering

File: ingestion/src/metadata/cmd.py [hunk 5]
```diff
@@ -282,6 +282,13 @@ def docker(
     default=None,
     required=False,
 )
+@click.option(
+    "--uploadDestinationType",
+    help="AWS or AZURE",
+    type=click.Choice(['AZURE', 'AWS'], case_sensitive=False),
```
Comment (pmbrull) [inline ingestion/src/metadata/cmd.py:288]: If we are not setting case sensitive here, the checks later (e.g., `uploadDestinationType == "AWS":`) will fail

**Review by pmbrull — COMMENTED** [review]
Thanks for the PR, overall it looks good, left some comments if you have the time

File: ingestion/src/metadata/cli/backup.py (hunk 2 above)
Comment (amymareemc) [reply ingestion/src/metadata/cli/backup.py:137]: Encountering issue when running `make py_format`. Have asked on slack. Observed on sonar cloud that the problem was with variable names so have fixed this issue.

File: ingestion/src/metadata/cli/backup.py (hunk 4 above)
Comment (amymareemc) [reply ingestion/src/metadata/cli/backup.py]: Added Enum class in utils folder. Not sure if this is the appropriate location. Please could you review?

**Comment by amymareemc** [comment]
@pmbrull Changes have been made. Could you please review?

File: ingestion/src/metadata/cli/backup.py (hunk 4 above)
Comment (pmbrull) [reply ingestion/src/metadata/cli/backup.py]: Thanks for the changes @amymareemc!

A couple of notes here:
1. The upload destination type only impacts the backup module. We can create the Enum on `backup.py` directly after the imports.
2. Let's try to follow PEP8 naming conventions (https://peps.python.org/pep-0008/). Classes can be named in CamelCase, e.g., `UploadDestinationType`
3. This is just a personal preference, but I like Enums that express written constants to be strings, E.g.:

```python
class UploadDestinationType(Enum):
    AZURE = "Azure"

# Then you can compare as
if uploadDestinationType == UploadDestionationType.AZURE.value
```

This allows for string comparison of values without hardcoding anything

File: ingestion/src/metadata/cli/backup.py [hunk 6]
```diff
@@ -98,13 +100,63 @@ def upload_backup(endpoint: str, bucket: str, key: str, file: Path) -> None:
...
+    :param container: Azure container to upload file to
+    :param file: file to upload
+    """
+
+    try:
+        from azure.identity import DefaultAzureCredential
+
+        default_credential = DefaultAzureCredential()
+        # Create the BlobServiceClient object
+        blob_service_client = BlobServiceClient(
```
Comment (pmbrull) [inline ingestion/src/metadata/cli/backup.py:117]: Thanks for initializing

**Comment by pmbrull** [comment]
> @pmbrull Changes have been made. Could you please review?

Thanks @amymareemc for the changes, left a comment above

File: ingestion/src/metadata/cli/backup.py (hunk 4 above)
Comment (amymareemc) [reply ingestion/src/metadata/cli/backup.py]: Thanks so much for the notes, implementing these changes now!
Just on point 1, the Enum type is also used in cmd.py (see[ line 286](https://github.com/open-metadata/OpenMetadata/pull/8334/commits/084e6f8ebb431c0dfaf7e6197c7a853bec6551e9#diff-2b38722110cfd94ca38c333ca42feb6387e283888b5dda34081e7421c83fe925L286)). Given this, should it still just be created in `backup.py`?

Comment (pmbrull) [reply ingestion/src/metadata/cli/backup.py]: yess that should not be an issue, `cmd.py` can import the enum from `backup.py` directly

**Comment by pmbrull** [comment]
Hi @amymareemc, thanks for the changes. Were you able to test it out? If it works feel free to mark the PR as ready for review and we can merge this in 🚀

**Comment by amymareemc** [comment]
Hi @pmbrull thanks for the help. It looks good on my end so I've put it up for review

**Comment by pmbrull** [comment]
> Hi @pmbrull thanks for the help. It looks good on my end so I've put it up for review

Thanks @amymareemc, the cypress tests are flaky so that's fine. But could you please run `make py_format` to fix the python CI? Thanks

**Comment by amymareemc** [comment]
Hi @pmbrull I have already run `make py_format` and I just did it again but nothing has changed. Not sure why it isn't passing the python CI?

**Comment by pmbrull** [comment]
> 

oh there are some linting issues. 1 sec

File: ingestion/src/metadata/cli/backup.py [hunk 7]
```diff
@@ -98,13 +106,64 @@ def upload_backup(endpoint: str, bucket: str, key: str, file: Path) -> None:
...
+        logger.debug(traceback.format_exc())
+        logger.error("Revisit the values of --upload")
+        raise err
+    except Exception as err:
+        logger.debug(traceback.format_exc())
+        logger.error(err)
+        raise err
+
+
 def run_backup(  # pylint: disable=too-many-arguments
```
Comment (pmbrull) [inline ingestion/src/metadata/cli/backup.py:159]:
```suggestion
def run_backup(  # pylint: disable=too-many-arguments,too-many-locals
```

Comment (pmbrull) [reply ingestion/src/metadata/cli/backup.py:159]: we can fix that later, as we have a ticket to restructure how we are passing the parameters

**Review by pmbrull — COMMENTED** [review]
Hei @amymareemc left some comments for the linting.

also please, update the `setup.py` file (https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/setup.py#L71) to add the library needed for Azure backup

**Comment by amymareemc** [comment]
@pmbrull thanks for the help with the linting. I've implemented your suggestions.

**Review by pmbrull — APPROVED** [review]
Thanks for the PR!

**Comment by joserodriguezc** [comment]
Hi all! This feature will be in the next release? Thanks!!!

**Pull request merged**